Enable open partition on frame boundary

ycho opened the pull request:
Issue #2166

- AV1 spec allows to use a partition that straddle on frame boundary if
  certain condition met, i.e, 'Not Split' is allowed if more than half of
  the partition size is available (based on coded frame size, i.e. multipe
  of 8 pixels) in either horizontal or vertical.

- For top-down partition search (i.e current speed levels 2~10),
  use bottomup on frame boundary

- Functon for computing visible area of a partition or a tx block based
  on actual frame size (not a coded size) is added and in general it should
  be only used for a compute distortion. Hence, do not use it for adjusting the block size.
  One good critical example would be that a coefficient block size, when it is encoded, is not affected by visibility.

- CFL specially requires to read outside the coded frame, when it compute average
  and obtain ac components for luma block, i.e. luma_ac().

lu-zero commented on src/predict.rs:
```diff
@@ -668,18 +670,20 @@ pub(crate) mod rust {
     let avg = T::cast_from((sum + (width >> 1) as u32) / width as u32);
-    for line in output.rows_iter_mut().take(height) {
-      line[..width].iter_mut().for_each(|v| *v = avg);
+    for y in 0..height {
```
Do not do that, we should assemble the border tiles/regions so they are aware they have some extra room while the other tiles/regions do not.

ycho replied:
Btw, I believe you aware this rust function does not run if SIMD available, which does what I have changed. I.e, cross the region border following av1 spec.

ycho replied:
So, I was not able to figure out other quick feasible way to achieve what I needed.

ycho replied:
> Do not do that, we should assemble the border tiles/regions so they are aware they have some extra room while the other tiles/regions do not.

Could you think of or suggest some other better ways to solve this? I've thought that we can split the function to have two versions, one for non frame boundary, and another for boundary case. But this split would be required every function affected by this PR and seemed not a good solution to me.

ycho replied:
> Do not do that, we should assemble the border tiles/regions so they are aware they have some extra room while the other tiles/regions do not.

Oh, I now understand. by 'border' you meant frame boundary. Then, this basically fall into the solution that I think and wanted to avoid.

ycho replied:
@lu-zero, I think the concern you have raised (that is valid) is cleared now, given that the src/predict.rs is intact now.
Can you +1 then?

rzumer commented:
Can you please run at least one comparison on AWCY and post results?

ycho commented:
@barrbrain, when you have time, please review the changes in luma_ac().
https://github.com/xiph/rav1e/pull/2396/files#diff-26006a0497afb96d876927e114a7b29eR1929,
which handles the encoding cases that does not occur with current mater. I.e. multiple tx blocks in luma and some of them are outside the encoded frame (i.e. clipped by 8 pixels).

ycho commented:
> Can you please run at least one comparison on AWCY and post results?

Sure, 

> Can you please run at least one comparison on AWCY and post results?

Sure, in fact this PR tries to solve #2166, there progress and several awcy are posted.

[awcy, master vs this work, objective-1-fast, speed 0](https://beta.arewecompressedyet.com/?job=master_c6bf0bf_1f_s0%402020-05-20T23%3A17%3A10.299Z&job=test_open_partition_fix2_s0%402020-05-27T06%3A14%3A13.233Z)

rzumer commented:
Thanks- I hadn't kept up with the comments on #2166.

barrbrain commented:
The aarch64 tests are failing with an out-of-bounds write in `crate::predict::rust::pred_directional()`:
https://github.com/xiph/rav1e/blob/9e9b7d55b7a3637052070e0f35ff20f07cc9fb64/src/predict.rs#L1297

```
thread '<unnamed>' panicked at 'index out of bounds: the len is 40 but the index is 40', src/predict.rs:1297:11
test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d ... FAILED

thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
test test_encode_decode::speed_1_dav1d ... FAILED

thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
test test_encode_decode::speed_0_dav1d ... FAILED

thread '<unnamed>' panicked at 'index out of bounds: the len is 56 but the index is 56', src/predict.rs:1297:11
test test_encode_decode::odd_size_frame_with_full_rdo_dav1d ... FAILED
```

ycho commented:
> The aarch64 tests are failing with an out-of-bounds write in `crate::predict::rust::pred_directional()`:
> 
> https://github.com/xiph/rav1e/blob/9e9b7d55b7a3637052070e0f35ff20f07cc9fb64/src/predict.rs#L1297
> 
> ```
> thread '<unnamed>' panicked at 'index out of bounds: the len is 40 but the index is 40', src/predict.rs:1297:11
> test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d ... FAILED
> 
> thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> test test_encode_decode::speed_1_dav1d ... FAILED
> 
> thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> test test_encode_decode::speed_0_dav1d ... FAILED
> 
> thread '<unnamed>' panicked at 'index out of bounds: the len is 56 but the index is 56', src/predict.rs:1297:11
> test test_encode_decode::odd_size_frame_with_full_rdo_dav1d ... FAILED
> ```

Yes, thanks for checking this.
I am middle of where output parameter is sent to pred_directional( ) with its size bounded in PlaneRegion structure, which I expect will crash and panic rust. If planeRegion has only its origin and no size, it will be fine to my design.

ycho commented:
> > The aarch64 tests are failing with an out-of-bounds write in `crate::predict::rust::pred_directional()`:
> > https://github.com/xiph/rav1e/blob/9e9b7d55b7a3637052070e0f35ff20f07cc9fb64/src/predict.rs#L1297
> > 
> > ```
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 40 but the index is 40', src/predict.rs:1297:11
> > test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d ... FAILED
> > 
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> > test test_encode_decode::speed_1_dav1d ... FAILED
> > 
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> > test test_encode_decode::speed_0_dav1d ... FAILED
> > 
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 56 but the index is 56', src/predict.rs:1297:11
> > test test_encode_decode::odd_size_frame_with_full_rdo_dav1d ... FAILED
> > ```
> 
> Yes, thanks for checking this.
> I am middle of where output parameter is sent to pred_directional( ) with its size bounded in PlaneRegion structure, which I expect will crash and panic rust. If planeRegion has only its origin and no size, it will be fine to my design.

Also, strnage that it is not reproducible on my local computer, mac os.

ycho commented:
> > The aarch64 tests are failing with an out-of-bounds write in `crate::predict::rust::pred_directional()`:
> > https://github.com/xiph/rav1e/blob/9e9b7d55b7a3637052070e0f35ff20f07cc9fb64/src/predict.rs#L1297
> > 
> > ```
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 40 but the index is 40', src/predict.rs:1297:11
> > test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d ... FAILED
> > 
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> > test test_encode_decode::speed_1_dav1d ... FAILED
> > 
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> > test test_encode_decode::speed_0_dav1d ... FAILED
> > 
> > thread '<unnamed>' panicked at 'index out of bounds: the len is 56 but the index is 56', src/predict.rs:1297:11
> > test test_encode_decode::odd_size_frame_with_full_rdo_dav1d ... FAILED
> > ```
> 
> Yes, thanks for checking this.
> I am middle of where output parameter is sent to pred_directional( ) with its size bounded in PlaneRegion structure, which I expect will crash and panic rust. If planeRegion has only its origin and no size, it will be fine to my design.

I was able to run the failed test on my raspberry pi3, and I could not reproduce the errors from Travis.
The 22 tests has all succeeded as below.

pi@raspberrypi:~/workspace/rav1e $ cargo test --release --features=decode_test_dav1d --verbose --color=always -- --color=always --ignored
       Fresh autocfg v1.0.0
       Fresh cfg-if v0.1.10
       Fresh lazy_static v1.4.0
       Fresh unicode-xid v0.2.1
       Fresh regex-syntax v0.6.18
       Fresh semver-parser v0.7.0
       Fresh scopeguard v1.1.0
       Fresh either v1.5.3
       Fresh version_check v0.9.2
       Fresh glob v0.3.0
       Fresh unicode-width v0.1.7
       Fresh ansi_term v0.11.0
       Fresh vec_map v0.8.2
       Fresh quick-error v1.2.3
       Fresh strsim v0.8.0
       Fresh termcolor v1.1.0
       Fresh peeking_take_while v0.1.2
       Fresh error-chain v0.10.0
       Fresh rustc-hash v1.1.0
       Fresh lazycell v1.2.1
       Fresh pkg-config v0.3.17
       Fresh toml v0.2.1
       Fresh shlex v0.1.1
       Fresh proc-macro-hack v0.5.16
       Fresh noop_proc_macro v0.3.0
       Fresh arc-swap v0.4.7
       Fresh itoa v0.4.6
       Fresh lab v0.7.2
       Fresh bitstream-io v0.8.5
       Fresh difference v2.0.0
       Fresh y4m v0.6.0
       Fresh predicates-core v1.0.0
       Fresh scan_fmt v0.2.5
       Fresh arrayvec v0.5.1
       Fresh same-file v1.0.6
       Fresh treeline v0.1.0
       Fresh ppv-lite86 v0.2.8
       Fresh oorandom v11.1.2
       Fresh thread_local v1.0.1
       Fresh semver v0.9.0
       Fresh itertools v0.9.0
       Fresh textwrap v0.11.0
       Fresh humantime v1.3.0
       Fresh metadeps v1.1.2
       Fresh paste-impl v0.1.18
       Fresh ivf v0.1.0 (/home/pi/workspace/rav1e/ivf)
       Fresh pretty_assertions v0.6.1
       Fresh predicates v1.0.4
       Fresh walkdir v2.3.1
       Fresh predicates-tree v1.0.0
       Fresh libc v0.2.71
       Fresh memchr v2.3.3
       Fresh proc-macro2 v1.0.18
       Fresh bitflags v1.2.1
       Fresh maybe-uninit v2.0.0
       Fresh rustc_version v0.2.3
       Fresh log v0.4.8
       Fresh serde v1.0.114
       Fresh ryu v1.0.5
       Fresh byteorder v1.3.4
       Fresh paste v0.1.18
       Fresh doc-comment v0.3.3
       Fresh num-traits v0.2.12
       Fresh jobserver v0.1.21
       Fresh crossbeam-utils v0.7.2
       Fresh num_cpus v1.13.0
       Fresh atty v0.2.14
       Fresh time v0.1.43
       Fresh which v3.1.1
       Fresh termios v0.3.2
       Fresh signal-hook-registry v1.2.0
       Fresh terminal_size v0.1.12
       Fresh getrandom v0.1.14
       Fresh wait-timeout v0.2.0
       Fresh memoffset v0.5.4
       Fresh quote v1.0.7
       Fresh aho-corasick v0.7.13
       Fresh csv-core v0.1.10
       Fresh fern v0.6.0
       Fresh nom v5.1.2
       Fresh serde_json v1.0.55
       Fresh regex-automata v0.1.9
       Fresh cc v1.0.54
       Fresh num-integer v0.1.43
       Fresh crossbeam-queue v0.2.3
       Fresh plotters v0.2.15
       Fresh clap v2.33.1
       Fresh signal-hook v0.1.16
       Fresh rand_core v0.5.1
       Fresh assert_cmd v1.0.1
       Fresh regex v1.3.9
       Fresh syn v1.0.33
       Fresh crossbeam-epoch v0.8.2
       Fresh simd_helpers v0.1.0
       Fresh cexpr v0.4.0
       Fresh tinytemplate v1.1.0
       Fresh chrono v0.4.11
       Fresh bstr v0.2.13
       Fresh rand_chacha v0.2.2
       Fresh env_logger v0.7.1
       Fresh console v0.11.3
       Fresh crossbeam-deque v0.7.3
       Fresh thiserror-impl v1.0.20
       Fresh num-derive v0.3.0
       Fresh arg_enum_proc_macro v0.3.0
       Fresh serde_derive v1.0.114
       Fresh interpolate_name v0.2.3
       Fresh cast v0.2.3
       Fresh csv v1.1.3
       Fresh rand v0.7.3
       Fresh rayon-core v1.7.1
       Fresh thiserror v1.0.20
       Fresh v_frame v0.1.0 (/home/pi/workspace/rav1e/v_frame)
       Fresh libloading v0.5.2
       Fresh criterion-plot v0.4.2
       Fresh rayon v1.3.1
       Fresh av-metrics v0.5.1
       Fresh clang-sys v0.29.3
       Fresh vergen v3.0.4 (/home/pi/workspace/rav1e/crates/vergen)
       Fresh bindgen v0.54.0
       Fresh nasm-rs v0.1.7
       Fresh criterion v0.3.2
       Fresh dav1d-sys v0.3.2
       Fresh rav1e v0.3.0 (/home/pi/workspace/rav1e)
    Finished release [optimized + debuginfo] target(s) in 2.08s
     Running `/home/pi/workspace/rav1e/target/release/deps/rav1e-def2548c0a7e3ff0 --color=always --ignored`

running 22 tests
test test_encode_decode::bitrate_dav1d ... ok
test test_encode_decode::chroma_sampling_422_dav1d ... ok
test test_encode_decode::chroma_sampling_400_dav1d ... ok
test test_encode_decode::chroma_sampling_420_dav1d ... test test_encode_decode::chroma_sampling_420_dav1d has been running for over 60 seconds
test test_encode_decode::chroma_sampling_444_dav1d ... test test_encode_decode::chroma_sampling_444_dav1d has been running for over 60 seconds
test test_encode_decode::high_bit_depth_10_dav1d ... test test_encode_decode::high_bit_depth_10_dav1d has been running for over 60 seconds
test test_encode_decode::chroma_sampling_420_dav1d ... ok
test test_encode_decode::chroma_sampling_444_dav1d ... ok
test test_encode_decode::high_bit_depth_10_dav1d ... ok
test test_encode_decode::high_bit_depth_12_dav1d ... test test_encode_decode::high_bit_depth_12_dav1d has been running for over 60 seconds
test test_encode_decode::high_bit_depth_12_dav1d ... ok
test test_encode_decode::low_bit_depth_dav1d ... test test_encode_decode::low_bit_depth_dav1d has been running for over 60 seconds
test test_encode_decode::odd_size_frame_with_full_rdo_dav1d ... test test_encode_decode::odd_size_frame_with_full_rdo_dav1d has been running for over 60 seconds
test test_encode_decode::rdo_loop_decision_cdef_sanity_dav1d ... test test_encode_decode::rdo_loop_decision_cdef_sanity_dav1d has been running for over 60 seconds
test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d ... test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d has been running for over 60 seconds
test test_encode_decode::low_bit_depth_dav1d ... ok
test test_encode_decode::rdo_loop_decision_cdef_sanity_dav1d ... ok
test test_encode_decode::speed_10_dav1d ... ok
test test_encode_decode::speed_0_dav1d ... test test_encode_decode::speed_0_dav1d has been running for over 60 seconds
test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d ... ok
test test_encode_decode::speed_2_dav1d ... ok
test test_encode_decode::speed_1_dav1d ... test test_encode_decode::speed_1_dav1d has been running for over 60 seconds
test test_encode_decode::speed_3_dav1d ... ok
test test_encode_decode::speed_1_dav1d ... ok
test test_encode_decode::speed_4_dav1d ... ok
test test_encode_decode::speed_6_dav1d ... ok
test test_encode_decode::speed_5_dav1d ... ok
test test_encode_decode::speed_7_dav1d ... ok
test test_encode_decode::speed_8_dav1d ... ok
test test_encode_decode::speed_9_dav1d ... ok
test test_encode_decode::speed_0_dav1d ... ok

test test_encode_decode::odd_size_frame_with_full_rdo_dav1d ... ok

test result: ok. 22 passed; 0 failed; 0 ignored; 0 measured; 134 filtered out

     Running `/home/pi/workspace/rav1e/target/release/deps/rav1e-758a4325eda9f75f --color=always --ignored`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running `/home/pi/workspace/rav1e/target/release/deps/binary-2140c47f49bb8a10 --color=always --ignored`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out

   Doc-tests rav1e
     Running `rustdoc --edition=2018 --crate-type lib --test /home/pi/workspace/rav1e/src/lib.rs --crate-name rav1e -L dependency=/home/pi/workspace/rav1e/target/release/deps -L native=/home/pi/workspace/rav1e/target/release/build/libloading-dc6eaaee7161ae07/out -L native=/usr/local/lib/arm-linux-gnueabihf -L dependency=/home/pi/workspace/rav1e/target/release/deps --test-args --color=always --test-args --ignored --cfg 'feature="clap"' --cfg 'feature="default"' --cfg 'feature="signal_support"' --cfg 'feature="av-metrics"' --cfg 'feature="regex"' --cfg 'feature="fern"' --cfg 'feature="scan_fmt"' --cfg 'feature="asm"' --cfg 'feature="cc"' --cfg 'feature="decode_test_dav1d"' --cfg 'feature="console"' --cfg 'feature="nasm-rs"' --cfg 'feature="signal-hook"' --cfg 'feature="binaries"' --cfg 'feature="ivf"' --cfg 'feature="y4m"' --cfg 'feature="dav1d-sys"' --extern arg_enum_proc_macro=/home/pi/workspace/rav1e/target/release/deps/libarg_enum_proc_macro-541232cffddfba23.so --extern arrayvec=/home/pi/workspace/rav1e/target/release/deps/libarrayvec-bec98196caa0b82e.rlib --extern assert_cmd=/home/pi/workspace/rav1e/target/release/deps/libassert_cmd-7b588dba0b2d25bd.rlib --extern av_metrics=/home/pi/workspace/rav1e/target/release/deps/libav_metrics-84265a1bdb2700ac.rlib --extern bitstream_io=/home/pi/workspace/rav1e/target/release/deps/libbitstream_io-671c6e4257a3a640.rlib --extern cfg_if=/home/pi/workspace/rav1e/target/release/deps/libcfg_if-4937682f2fc3d239.rlib --extern clap=/home/pi/workspace/rav1e/target/release/deps/libclap-a3f957d52076b194.rlib --extern console=/home/pi/workspace/rav1e/target/release/deps/libconsole-a3792d939f2737b5.rlib --extern criterion=/home/pi/workspace/rav1e/target/release/deps/libcriterion-9a28fe55b2c35695.rlib --extern dav1d_sys=/home/pi/workspace/rav1e/target/release/deps/libdav1d_sys-6c5ff34184c0ae6e.rlib --extern fern=/home/pi/workspace/rav1e/target/release/deps/libfern-c5e363e0d05befe5.rlib --extern interpolate_name=/home/pi/workspace/rav1e/target/release/deps/libinterpolate_name-d25a2055c1207420.so --extern itertools=/home/pi/workspace/rav1e/target/release/deps/libitertools-f6b5ef7158c021ac.rlib --extern ivf=/home/pi/workspace/rav1e/target/release/deps/libivf-68b37a981d09a085.rlib --extern libc=/home/pi/workspace/rav1e/target/release/deps/liblibc-3acae17f1de702b6.rlib --extern log=/home/pi/workspace/rav1e/target/release/deps/liblog-2dcd6041885011b1.rlib --extern noop_proc_macro=/home/pi/workspace/rav1e/target/release/deps/libnoop_proc_macro-a3e490dbd799dacb.so --extern num_derive=/home/pi/workspace/rav1e/target/release/deps/libnum_derive-f3e5755c57193155.so --extern num_traits=/home/pi/workspace/rav1e/target/release/deps/libnum_traits-e2379d93cc767813.rlib --extern paste=/home/pi/workspace/rav1e/target/release/deps/libpaste-96699211c365e50f.rlib --extern pretty_assertions=/home/pi/workspace/rav1e/target/release/deps/libpretty_assertions-d8e550af1e835b04.rlib --extern rand=/home/pi/workspace/rav1e/target/release/deps/librand-6b0e3d8f992b2309.rlib --extern rand_chacha=/home/pi/workspace/rav1e/target/release/deps/librand_chacha-e4189061907ca944.rlib --extern rav1e=/home/pi/workspace/rav1e/target/release/deps/librav1e-cd21268f22c8bf59.rlib --extern rayon=/home/pi/workspace/rav1e/target/release/deps/librayon-db23dff61fce2d17.rlib --extern scan_fmt=/home/pi/workspace/rav1e/target/release/deps/libscan_fmt-f9f7030042c4b09d.rlib --extern semver=/home/pi/workspace/rav1e/target/release/deps/libsemver-0697714599cb2e51.rlib --extern signal_hook=/home/pi/workspace/rav1e/target/release/deps/libsignal_hook-0f18c39d819f203c.rlib --extern simd_helpers=/home/pi/workspace/rav1e/target/release/deps/libsimd_helpers-3e4356ac48ebb273.so --extern thiserror=/home/pi/workspace/rav1e/target/release/deps/libthiserror-602c18181966beff.rlib --extern v_frame=/home/pi/workspace/rav1e/target/release/deps/libv_frame-3d4331777b397286.rlib --extern y4m=/home/pi/workspace/rav1e/target/release/deps/liby4m-66f323206f6f7309.rlib`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 6 filtered out

pi@raspberrypi:~/workspace/rav1e $

ycho commented:
> The aarch64 tests are failing with an out-of-bounds write in `crate::predict::rust::pred_directional()`:
> 
> https://github.com/xiph/rav1e/blob/9e9b7d55b7a3637052070e0f35ff20f07cc9fb64/src/predict.rs#L1297
> 
> ```
> thread '<unnamed>' panicked at 'index out of bounds: the len is 40 but the index is 40', src/predict.rs:1297:11
> test test_encode_decode::rdo_loop_decision_lrf_sanity_dav1d ... FAILED
> 
> thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> test test_encode_decode::speed_1_dav1d ... FAILED
> 
> thread '<unnamed>' panicked at 'index out of bounds: the len is 8 but the index is 8', src/predict.rs:1297:11
> test test_encode_decode::speed_0_dav1d ... FAILED
> 
> thread '<unnamed>' panicked at 'index out of bounds: the len is 56 but the index is 56', src/predict.rs:1297:11
> test test_encode_decode::odd_size_frame_with_full_rdo_dav1d ... FAILED
> ```

fiy, i confirm that this class of errors will happen when rust functions that access memory trying to cross left or bottom coded (i.e. multiple of 8 pixels) frame boundaries. More specifically, when trying to access memory that are out of rectangular bounds defined in TileStatMut via TileRect, PlaneRegion and/or subregion().

And the reason why it happens for arm target only with Travis is that there is no arm NEON code for pred_directional( ), thus rust functions are called for it and error triggered.

ycho commented:
> on running native-run(ie. without having assembly functions), tests are failing,
> 
> Command used: `env RUST_BACKTRACE=full RAV1E_CPU_TARGET=rust cargo test --release --features=decode_test_dav1d --color=always -- --color=always --ignored`
> 
> Short Output: `test result: FAILED. 9 passed; 13 failed; 0 ignored; 0 measured; 400 filtered out`
> 
> Full log: https://gist.github.com/vibhoothiiaanand/1e6fbbd6bc8dea27e5cfa4b3bdc07274

Wow- thanks for trying this. Yeah, this is what you have asked during watercooler today. I admit that it fails now much more than last time I did. I think since I've fixed some panics I met then I did not test with 'RAV1E_CPU_TARGET=rust' again with much more work as added to enable open partition.

So, as I stated in the watercooler, I plan to revert all changes I replaced for row_iter_[mut] based prediction and other functions once I implement better way of keep using current PlaneRegion and its bounding box memory access enforcement design.

ycho commented:
I've rebased this PR based on my other PR #2445, since this PR is dependent on it.

ycho commented:
Cannot locally reproduce the failed test of no-asm test,
https://github.com/xiph/rav1e/pull/2396/checks?check_run_id=896459487#step:20:467

I used the command (on Mac OS), 
_export RAV1E_CPU_TARGET=rust; cargo test --release --features=decode_test_dav1d --verbose --color=always -- --ignored_
and 
_export RAV1E_CPU_TARGET=rust; cargo test --release --features=decode_test_dav1d --verbose --color=always_

ycho commented:
> I've rebased this PR based on my other PR #2445, since this PR is dependent on it.

Done.

ycho commented:
> Cannot locally reproduce the failed test of no-asm test,
> https://github.com/xiph/rav1e/pull/2396/checks?check_run_id=896459487#step:20:467
> 
> I used the command (on Mac OS),
> _export RAV1E_CPU_TARGET=rust; cargo test --release --features=decode_test_dav1d --verbose --color=always -- --ignored_
> and
> _export RAV1E_CPU_TARGET=rust; cargo test --release --features=decode_test_dav1d --verbose --color=always_

@vibhoothiiaanand , sorry to ask this, with the same test command you informed me last tine (and mac os version is shown above), could you check whether you can see/reproduce the same test failure (only one, 'no-asm' test) from CI please?

ycho commented:
New awcy result after rebasing on ground work.
[awcy -s 0, low latency](https://beta.arewecompressedyet.com/?job=master_c89bf8bd_s0_low_latency%402020-07-21T09%3A18%3A05.565Z&job=open_partition_c89bf8bd_s0_low_latency%402020-07-22T21%3A47%3A47.679Z)


|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -2.0825 | -2.6507 | -2.6004 |  -2.0535 | -2.1375 | -2.1537 |    -2.5300 |

[awcy -s 6, low latency](https://beta.arewecompressedyet.com/?job=master_c89bf8bd_s6_low_latency%402020-07-21T08%3A39%3A47.175Z&job=open_partition_c89bf8bd_s6_low_latency%402020-07-22T21%3A46%3A46.250Z)

(Note: open-partition uses bottomup partition search if SB straddles on frame boundaries for -s 2~10 (i.e. topdown search is used as default), which itself on master gives 0.16% gain.

|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.5908 | -2.5491 | -2.4333 |  -1.5583 | -1.6250 | -1.6096 |    -1.9550 |

[awcy -s2, low latency](https://beta.arewecompressedyet.com/?job=master_c89bf8bd_s2_low_latency%402020-07-21T08%3A40%3A20.377Z&job=open_partition_c89bf8bd_s2_low_latency%402020-07-22T21%3A46%3A28.215Z)


|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.5974 | -2.0216 | -2.1389 |  -1.6513 | -1.6941 | -1.6904 |    -1.9199 |

ycho commented on src/encoder.rs:
```diff
@@ -3473,8 +3546,16 @@ fn encode_tile<'a, T: Pixel>(
       // Encode SuperBlock
-      if fi.config.speed_settings.encode_bottomup {
+      if fi.config.speed_settings.encode_bottomup
+        || is_straddle_sbx
```
What this change does is: If SB straddle on frame boundary, use bottom-up partition search because topdown cannot use open-partition due to its limited search depth of one level.
However, currently I observe this change CAUSE DESYNC, i.e. mismatch between the decoded video from encoder and decoder, so I think I need to fix this before landing.

ycho commented:
@KyleSiefring , https://github.com/xiph/rav1e/pull/2396/commits/9e06b16e76cce5ed71c1f6b4e9e5ac98f4c1b9fb has fixed the issue with your EPZS repair + open-parition
The commit is now placed at 1st commit of this PR.

ycho commented:
In order to use open-partition with top-down search, i.e speed 2~10, I need to also use bottomup for frame boundary, because otherwise just largest partition will be chosen with topdown there and gives a coding loss (~1.5%).
 
Using bottomup partition search together with inter-tx split seems causing desync under certain conditions.
Previously, I have tested inter-tx split for smaller min block size like 32x32, 16x16, and 8x8 at speed 9 or 10,
but I think I have not tested inter-tx split with bottomup, which is only used for speed 0 and 1.
Until this is fixed, I'd want to hold landing this PR.

ycho commented:
After fixing the inter-tx split for bottomup partition search, awcy results for speed 9 and 10 are available.

[awcy -s 10](https://beta.arewecompressedyet.com/?job=open_partition_c89bf8bd_s10_low_latency%402020-08-09T22%3A36%3A05.420Z&job=master_c89bf8bd_s10_low_latency%402020-07-23T06%3A13%3A14.540Z)


|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.2429 | -2.0600 | -2.0934 |  -1.3338 | -1.1566 | -1.2524 |    -1.5233 |

[awcy -s 9](https://beta.arewecompressedyet.com/?job=open_partition_c89bf8bd_s9_low_latency%402020-08-09T22%3A35%3A56.238Z&job=master_c89bf8bd_s9_low_latency%402020-07-21T08%3A40%3A31.196Z)



|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -1.6695 | -2.1247 | -2.1208 |  -1.7131 | -1.7301 | -1.7104 |    -2.0794 |

ycho commented:
[newer result with open partition enabled on awcy, 2.0% gain at default speed 6](https://beta.arewecompressedyet.com/?job=master-fe61a877488da642ae8c0844d31ff62f8b2ba66e&job=master-a83f57a90583dbd3b0db70398825c606e28e9425)

lu-zero commented:
That's great !
